design-proposals/klip-57-kafka-headers.md

```diff
@@ -0,0 +1,138 @@
+# KLIP 57 - Reading Kafka headers in ksqlDB
+**Author**: Zara Lim (@jzaralim) | 
+**Release Target**: 0.25 | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/8293
+
+**tl;dr:** _Expose Kafka header data in ksqlDB. Currently, headers are the only metadata in a Kafka record that is invisible to users._
+
+## Motivation and background
+
+So far, ksqlDB exposes the timestamp, partition and offset of Kafka records as pseudocolumns. The only metadata left that is invisible to ksqlDB users are [headers](https://kafka.apache.org/28/javadoc/org/apache/kafka/common/header/Headers.html). Kafka headers are a list of zero or more key-value pairs, where the keys are (not necessarily unique) strings and the values are byte arrays. Headers typically contain metadata about records, which can then be used for routing or processing (e.g. the header could store information about how to deserialize strings in the value). If ksqlDB exposes headers to the user, then these steps could be done in ksqlDB.
+
+## What is in scope
+* Syntax for accessing headers
+* Support for all query types except for `INSERT` queries
+* Including header columns in a data source's schema
+* Functions to help decode byte strings to primitive types (BYTES to STRING already exists, new functions would be conversions to INT, BIGINT and DOUBLE)
+
+## What is not in scope
+* Writing headers to output topics - we might do this in the future, but that is a separate discussion
+* Advanced functions to help decode byte strings (e.g. functions to deserialize complex serialized data.) - if a user wants to convert something that we don't have a function for, then they can either create their own UDF or use `CREATE FUNCTION` when that becomes available.
+
+## Public APIS
+
+### Syntax
+
+Unlike other pseudocolumns, if a user wants to access headers, they will create a column with the `HEADERS` keyword. This column will be populated with the full list of header keys and values.
+```
+ksql> CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS);
```

Comment:
Can I use `HEADERS` on different columns?

```
CREATE STREAM A (
  my_headers1 ARRAY<STRUCT<key STRING, value BYTES>> HEADERS,
  my_headers2 ARRAY<STRUCT<key STRING, value BYTES>> HEADERS
);
```

Reply:
Yes, added to klip

Reply:
As discussed below, we should not allow it, but restrict a schema definition to at most one HEADER column.